### PR TITLE
Allow templates to follow named schedules

### LINKED ISSUE #543: [Feature Request] Budget indication of scheduled transactions
### Discussed in https://github.com/actualbudget/actual/discussions/177

<div type='discussions-op-text'>

<sup>Originally posted by **alpha-beta-soup** May  8, 2022</sup>
Let's say I pay $500 rent every Friday. Some months there are four Fridays in a month, less often there are five. Do I budget $2000 or $2500 in a given month?

Let's also say I have a complex series of subscriptions, tithes, donations, fees, etc. that repeat either weekly, fortnightly, monthy, 6-monthly, annually, bi-annually... but definitely _deterministically_. How much do I budget in these categories?

I should be able to tell, within a budget category for any month, what all of my _scheduled expenses_ are, so I can budget for these known and regular expenses before allocating money to categories to cover my unknown/irregular expenses.

Although I'm new to Actual (recent nYNAB convert due to their massive price hike), I can't see a way to see these expenses reflected in my budget _until after they occur_. I want to budget for them _before they occur_.</div>

### PR DESCRIPTION
**Pull request by pole95**
This PR:
- Introduces a new `schedule` keyword for templates, which will put as much money in the category as is needed to fill any payments of that schedule in the current month
- Allows schedules to optionally be named, in order to reference them by name in the template

All in all, this schould fix #543. The schedules become the source of truth for assigning the budget, and any changes in the schedules are reflected as soon as you update the template.

### CODE REVIEW

**Comment by shall0pass** [comment]
This is an exciting addition!

I tried it but it's giving an unexpected result.  The schedule I created was for April 30 but the funding target isn't complete until the month of May.

template_schedule

I also can't use this template further out in the budget for future months.  That may be by design or just a fact that specific schedule doesn't exist in the scheduler yet because the prior schedule hasn't been completed.

**Comment by pole95** [comment]
Thanks for testing, I'll look into why its splitting that up into two months.

The second part is partly due to design, prior balance in the category is taken into account. Since the $50 are not yet spent, no money will be added to the category. I'm not sure which way would be best in handling this, so I am open to suggestions (maybe we could add a switch to the template definition if prior balance should be considered?)

**Comment by pole95** [comment]
@shall0pass Unfortunately I cannot reproduce the bug you are experiencing, neither by setting the schedule to a specific day (like in your example on the 30th of each month), nor by setting it the last day of the month.
Could you try reproducing it on your end and provide an exact step by step on how you get this behavior to show up?

**Comment by shall0pass** [comment]
I started with a fresh budget (no accounts, no transactions), added an account, created 1 schedule, then created 1 template.  I can make a screen gif again if that'll be helpful from beginning to end.

**Comment by shall0pass** [comment]
I used the netlify preview build to record this. I pushed the date out 1 month to the end of May.  I expect to see two equal amounts (April and May) to meet the schedule, but it looks like it is trying to split it into 3 amounts.  If I only apply the template for the month of May, and not April, it budgets half the amount.  It looks like is is calculating to (amount needed)/(months + 1) instead of (amount needed)/(months).

schedule_template

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 1]
```diff
@@ -333,6 +350,42 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
+        let conditions = rule.serialize().conditions;
+        let { date: dateCond, amount: amountCond } =
+          extractScheduleConds(conditions);
+        let next_date_string = getNextDate(dateCond, current_month);
+        let num_months = differenceInCalendarMonths(
+          new Date(next_date_string),
+          current_month,
+        );
+        let target = -getScheduledAmount(amountCond.value);
+        let diff = target - last_month_balance;
```
Comment (shall0pass) [inline packages/loot-core/src/server/budget/goaltemplates.js:368]: I think this should probably reflect "this months" balance.  There could potentially be spending present for the current month that wouldn't be taken into account.

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 2]
```diff
@@ -333,6 +350,42 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
...
+        if (num_months > 0) {
+          if (diff >= 0 && num_months > -1) {
+            to_budget += Math.round(diff / (num_months + 1));
```
Comment (shall0pass) [inline packages/loot-core/src/server/budget/goaltemplates.js:371]: This is where the the monthly allocation error seems to be that I'm seeing.
```suggestion
            to_budget += Math.round(diff / num_months);
```

**Comment by shall0pass** [comment]
I also tried a different scenario.  This time I put an existing transaction in the month, so I had a transaction for the same month I am applying the template.  I think maybe we should use the current category balance instead of last months category balance like we do on others.  The template didn't add enough to cover the upcoming expense in addition to the spending that already occurred.  I hope this screen grab will demonstrate what I mean.

schedule_template2

File: packages/loot-core/src/server/budget/goaltemplates.js (hunk 1 above)
Comment (shall0pass) [reply packages/loot-core/src/server/budget/goaltemplates.js:368]: How would you feel about this?  It uses this months balance and the already budgeted amount to determine what's needed.
```suggestion
        let diff = target - balance + budgeted;
```

Comment (pole95) [reply packages/loot-core/src/server/budget/goaltemplates.js:368]: Sounds good, this is probably more in line what a "layman" would expect the template to do.

File: packages/loot-core/src/server/budget/goaltemplates.js (hunk 2 above)
Comment (pole95) [reply packages/loot-core/src/server/budget/goaltemplates.js:371]: Yes, this should probably fix it. Brainless copy-paste gone wrong, although I still don't quite get why I could not replicate that error

**Comment by shall0pass** [comment]
I found one more issue.  If you have two schedule templates in the same category (I'm assuming that's a feature we want), scheduled 1 month away, the first month calculates properly but the second month does not.

**Comment by shall0pass** [comment]
Maybe one more issue.  If we have an schedule rule that references an old schedule that is non repeating, but the template line still exists in the notes, we get stuck in the while() loop and have to refresh the page to escape.  Maybe we could return an error if the loop count goes above a certain number of iterations?  Or, if we can determine if the schedule is non repeating and in the past to return an error explaining the issue?

Reproduce by:
1. Create a schedule in the past that is not repeating.
2. Add a template string that calls upon that schedule.
3. Go at least 1 month past the missed schedule and apply the templates.

**Comment by pole95** [comment]
Thanks for your additional comments. It shows how little I thought about use cases outside of my very specific one. I should be able to get the fixes in on the weekend.

**Comment by j-f1** [comment]
Also a heads up — I’m not entirely on board with this change UI-wise. I tried to add a name field to schedules before (see #268) and part of the reason I gave up was that I couldn’t find a good place to put the text field to edit it. One possibility could be to do a bigger redesign of the dialog to have two columns for the top part — one with name/payee/account/amount and another with date/repeats/auto-add/future dates. I haven’t actually tried that so I’m not sure it would work.

I would also like to see a checkbox to add the schedule name to the note column of scheduled transactions (alternatively, this could be implemented as a dynamic placeholder sort of thing? Once this merges I can open an issue to discuss this a bit more and come up with a solution)

**Comment by pole95** [comment]
The UI is certainly not perfect, but IMO this should be addressed in a different PR as it is a different problem. My changes to the UI are purely to allow the template feature to work, I'm open to any improvements there. However, unless there are significant problems with the current implementation, I wont change the UI anymore in this PR, as it offers the needed functionality to make the schedule templates work, while not looking completely hideous.

The same applies to your second point. It is not really related to my original feature of allowing schedules to be referenced as templates, so I would also suggest tackling this in a separate PR.

File: packages/loot-core/src/server/schedules/app.js [hunk 3]
```diff
@@ -176,6 +176,15 @@ export async function setNextDate({ id, start, conditions, reset }) {
 
 // Methods
 
+export async function checkIfScheduleExists(name, scheduleId) {
+  let idForName = await db.first('SELECT id from schedules WHERE name = ?', [
+    name,
+  ]);
+  return (
+    idForName !== null && (scheduleId ? idForName['id'] !== scheduleId : true)
+  );
```
Comment (j-f1) [inline packages/loot-core/src/server/schedules/app.js:185]: Tried to make this a little easier to follow, let me know what you think:

```suggestion
  let idForName = await db.first('SELECT id from schedules WHERE name = ?', [
    name,
  ]);

  if (idForName == null) {
    return false;
  }
  if (scheduleId) {
    return idForName['id'] !== scheduleId;
  }
  return true;
```

File: packages/loot-core/src/server/budget/goaltemplates.js [hunk 4]
```diff
@@ -166,6 +176,14 @@ async function applyCategoryTemplate(category, template_lines, month, force) {
           template.from = format(spend_from, 'yyyy-MM');
         }
         break;
+      case 'schedule':
+        if (!all_schedule_names.includes(template.name)) {
+          console.log(
+            `${category.name}: Schedule ${template.name} does not exist`,
+          );
```
Comment (j-f1) [inline packages/loot-core/src/server/budget/goaltemplates.js:183]: Can you add this to the `errors` array like other problems are reported in this function?

**Comment by j-f1** [comment]
Opened #908 & #909 to track my earlier comments.

File: packages/loot-core/src/server/budget/goaltemplates.js (hunk 4 above)
Comment (pole95) [reply packages/loot-core/src/server/budget/goaltemplates.js]: Oops, I missed that when I merged in the new error-reporting.

**Review by j-f1 — APPROVED** [review]
Looks good to me! There are some failures in the end-to-end test but I’m re-running since they seem to be unrelated to these changes.

**Comment by j-f1** [comment]
Hm, looks like the end-to-end tests are still failing. Can you take a look?

**Comment by pole95** [comment]
Looked like it was the update notification that confused playwright, hopefully the tests pass now

**Review by j-f1 — APPROVED** [review]
Great!

**Comment by j-f1** [comment]
Thank you for contributing!

**Pull request merged**